File: frameworks/runtime-src/Classes/app/FieldAuth.cpp
#include "FieldAuth.h"
#include "cocos2d.h"

using namespace cocos2d;

FieldAuth::FieldAuth()
{
}

FieldAuth::~FieldAuth()
{
    sqlite3_close(useDataBase);
}

void FieldAuth::initDatabase()
{
    CCLOG("FieldAuth::initDatabase start");
    
    //DBファイルの保存先のパス
    auto filePath = FileUtils::getInstance()->getWritablePath();
    filePath.append("PtouchDB.db");
    
    //OPEN
    auto status = sqlite3_open(filePath.c_str(), &useDataBase);
    
    //ステータスが0以外の場合はエラーを表示
    if (status != SQLITE_OK){
        CCLOG("open failed : %s", errorMessage);
    }else{
        CCLOG("open sucessed");
    }

    this->createTable();
    this->initData();
    
    CCLOG("FieldAuth::initDatabase end");
}

void FieldAuth::exec_sql(char *sql)
{
    auto status = sqlite3_exec(useDataBase, sql, NULL, NULL, &errorMessage );
    if( status != SQLITE_OK ) CCLOG("exec_sql failed : %s", errorMessage);
}

void FieldAuth::createTable()
{
    this->exec_sql((char *)"CREATE TABLE field_auth(id integer primary key autoincrement,\
                                                        field_name VARCHAR(32) UNIQUE,\
                                                        status VARCHAR(32))");
}

// 初期データの投入
// 2回目移行はDUPLICATEで更新されないはず
void FieldAuth::initData()
{
    this->exec_sql((char *)"INSERT INTO field_auth(field_name, status)\
                                            VALUES('animal', 'open')");
    this->exec_sql((char *)"INSERT INTO field_auth(field_name, status)\
                                            VALUES('vehicle', 'open')");
    this->exec_sql((char *)"INSERT INTO field_auth(field_name, status)\
                                            VALUES('vegetable', 'close')");
}

static int canEnterFieldCallback(void *data, int argc, char **argv, char **azColName){
//    int i;
//    for(i=0; i<argc; i++){
//        printf("%s = %s\n", azColName[i], argv[i] ? argv[i] : "NULL");
//    }
    // ココに来たってことはレコードがあるってことなので
    bool *exist_record = (bool *)data;
    *exist_record = true;
    return 0;
}

/*
 * 特定のフィールドに入れる権利が有るか
 */
bool FieldAuth::canEnterField(const char * field_name)
{
    char *sql = sqlite3_mprintf( "SELECT * FROM field_auth WHERE field_name == '%q' AND status=='open';", field_name );
    CCLOG( "%s\n", sql );
    
    bool exist_record = false;
    auto status = sqlite3_exec(useDataBase, sql, canEnterFieldCallback, &exist_record, &errorMessage);
    if( status != SQLITE_OK ) CCLOG("exec_sql failed : %s", errorMessage);
    
    sqlite3_free( sql );
    return exist_record;
}

/*
 * 特定のフィールドに入れる権利を付与
 */
void FieldAuth::enableEnterField(const char * field_name)
{
    char *sql = sqlite3_mprintf( "UPDATE field_auth SET status='open' WHERE field_name == '%q';", field_name );
    CCLOG( "%s\n", sql );
    
    this->exec_sql(sql);
    
    sqlite3_free( sql );

}

File: src/model/field_auth.js

var FIELD_INIT_DATA = [
    {
        field_name: 'animal',
        default_status: 'open',
        can_open_by_payment: false,
        can_open_by_share: false
    },
    {
        field_name: 'color',
        default_status: 'close',
        can_open_by_payment: false,
        can_open_by_share: true
    },
    {
        field_name: 'vegetable',
        default_status: 'close',
        can_open_by_payment: true,
        can_open_by_share: true
    },
    {
        field_name: 'vehicle',
        default_status: 'close',
        can_open_by_payment: true,
        can_open_by_share: false
    },
    {
        field_name: 'hoge',
        default_status: 'close',
        can_open_by_payment: true,
        can_open_by_share: false
    }
];

var FieldAuth = {
    _createTable:function () {
        var sql = "CREATE TABLE field_auth (" +
                    "field_name VARCHAR(32) UNIQUE," +
                    "status VARCHAR(32))";
        SQLTool.execSql("FieldAuth.db", sql);
    },
    _initData:function () {
        _.each(FIELD_INIT_DATA, function(element, index, array) {
            var sql = "INSERT INTO field_auth(" +
                            "field_name," +
                            "status)" +
                        "VALUES(" +
                            "'" + element.field_name + "'," +
                            "'" + element.default_status + "')";
            SQLTool.execSql("FieldAuth.db", sql);
        }, this);
    },
    init:function () {
        // 初期化してなければ初期化
        var sql = "SELECT count(*) as cnt FROM field_auth";
        var ret = SQLTool.execSql("FieldAuth.db", sql);
        if(ret.status === 0 && ret.result[0].cnt !== "0"){
            return;
        }
        FieldAuth._createTable();
        FieldAuth._initData();
    }
};

function Field() {
    /*
     * 特定のフィールドに入れる権利が有るか
     */
    this.canEnterField = function() {
        return FieldAuthAPI.canEnterField(this.field_name);
    };
};

var FieldAuthAPI = {
    /*
     * DB初期化
     */
    initDB:function () {
        FieldAuth.init();
    },

    canEnterField:function (field_name) {
        var sql = "SELECT count(*) as cnt FROM field_auth " +
                    "WHERE " +
                        "field_name == '" + field_name + "' AND " +
                        "status == 'open'";
        var ret = SQLTool.execSql("FieldAuth.db", sql);
        if(ret.status !== 0){
            cc.log("ERROR canEnterField!!!");
            return;
        }
        return Boolean(Number(ret.result[0].cnt));
    },

    /*
     * 特定のフィールドに入れる権利を付与
     */
    enableEnterField:function (field_name) {
        var accesser = new jsb.SQLiteBridge("FieldAuth.db");
        var sql = "UPDATE field_auth " + 
                        "SET status='open' " +
                    "WHERE " +
                        "field_name == '" + field_name + "'";
        var ret = SQLTool.execSql("FieldAuth.db", sql);
        if(ret.status !== 0){
            cc.log("ERROR enableEnterField!!!");
            return;
        }
    },

    /*
     * フィールド情報一覧
     */
    allFields:function () {
        return _.map(FIELD_INIT_DATA, function(element) {
            field = new Field();
            return _.extend(field, element);
        });
    }
};
